### DISCUSSION TITLE
Training Output Neuron to Spike at a Specified Frequency

### DISCUSSION THREAD

**Discussion by DS-567** [jeshraghian/snntorch#204] [General]
Hi, I have started using snntorch at it looks very effective!

When training output neurons using rate coding, am I correct in saying that the neuron with the highest firing frequency class wins?

Is it possible using snntorch to specify an output frequency that you want an output neuron to spike at? E.g. enter output frequency parameter in code for an output neuron to spike at (e.g. 100Hz) when it recognises an input pattern sample? So when this input pattern is inputted, the output neuron will spike at 100Hz. And if a different pattern is inputted, the output frequency of the same trained output neuron will then be different?

**Comment by jeshraghian** [comment]
Yep - this is the most common way to do it.

It's definitely possible using `snntorch.functional.mse_count_rate()` (docs [here](https://snntorch.readthedocs.io/en/latest/snntorch.functional.html#snntorch.functional.loss.mse_count_loss)).

You can set the total proportion of desires spikes by modifying the arguments:
* `correct_rate = 1` sets the firing rate to 100% of the total time steps of the target class
* `correct_rate = 0.8` would set the firing rate to 80% of the total time steps of the target class
* `incorrect_rate = 0` sets the firing rate to 0% of the total time steps of all non-target classes
* `incorrect_rate = 0.2` sets the firing rate to 20% of the total time steps of all non-target classes

How this translates into frequency is up to you and depends on what you define your time-step to be.

E.g., if dt = 1 ms, and you simulate 1000 steps (i.e., 1 second), and `correct_rate = 0.5`, then you're asking for a frequency of 500 Hz.

**Reply by DS-567** [reply]
Hi Jesh, thanks for your response. This sounds exactly like what we require. Do you mean the `mse_count_loss()` as I cannot find `mse_count_rate()` in the docs (link above)?

Do you know are there any example scripts showing how to most suitably implement this function in a rate based code training algorithm?
